
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Short and simple bioinformatics tutorials">
      
      
      
        <meta name="author" content="A. Sina Booeshaghi">
      
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.2.3">
    
    
      
        <title>Terminal interacting with fastq files - bytesizebio</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.3b61ea93.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.39b8e14a.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
      <link rel="stylesheet" href="../../css/custom.css">
    
    
      
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#interacting-with-fastq-files" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="../.." title="bytesizebio" class="md-header-nav__button md-logo" aria-label="bytesizebio">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      <div class="md-header-nav__ellipsis">
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            bytesizebio
          </span>
        </div>
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            
              Terminal interacting with fastq files
            
          </span>
        </div>
      </div>
    </div>
    
      <label class="md-header-nav__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" data-md-component="search-reset" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/sbooeshaghi/bytesizebio/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sbooeshaghi/bytesizebio
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="bytesizebio" class="md-nav__button md-logo" aria-label="bytesizebio">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    bytesizebio
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/sbooeshaghi/bytesizebio/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sbooeshaghi/bytesizebio
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      



  <li class="md-nav__item">
    <a href="../.." class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  


  
  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" checked>
    
    <label class="md-nav__link" for="nav-2">
      Tutorials
      <span class="md-nav__icon md-icon"></span>
    </label>
    <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        <span class="md-nav__icon md-icon"></span>
        Tutorials
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Terminal interacting with fastq files
        <span class="md-nav__icon md-icon"></span>
      </label>
    
    <a href="./" class="md-nav__link md-nav__link--active">
      Terminal interacting with fastq files
    </a>
    
      
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-curl-command" class="md-nav__link">
    The curl command
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#downloading-fastq-files" class="md-nav__link">
    Downloading FASTQ files
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#interacting-with-fastq-files_1" class="md-nav__link">
    Interacting with FASTQ files
  </a>
  
    <nav class="md-nav" aria-label="Interacting with FASTQ files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#useful-commands" class="md-nav__link">
    Useful commands
  </a>
  
    <nav class="md-nav" aria-label="Useful commands">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-gzip-command-and-the-gunzip-command" class="md-nav__link">
    The gzip command and the gunzip command
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-zcat-command" class="md-nav__link">
    The zcat command
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reading-fastq-files" class="md-nav__link">
    Reading FASTQ files
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fastq-format" class="md-nav__link">
    FASTQ Format
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#counting-the-number-of-molecules-reads" class="md-nav__link">
    Counting the number of molecules (reads)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#finding-an-exact-sequence" class="md-nav__link">
    Finding an exact sequence
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#enumerating-lines-of-a-fastq-file" class="md-nav__link">
    Enumerating lines of a FASTQ file
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary" class="md-nav__link">
    Summary
  </a>
  
</li>
      
    </ul>
  
</nav>
    
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-curl-command" class="md-nav__link">
    The curl command
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#downloading-fastq-files" class="md-nav__link">
    Downloading FASTQ files
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#interacting-with-fastq-files_1" class="md-nav__link">
    Interacting with FASTQ files
  </a>
  
    <nav class="md-nav" aria-label="Interacting with FASTQ files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#useful-commands" class="md-nav__link">
    Useful commands
  </a>
  
    <nav class="md-nav" aria-label="Useful commands">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-gzip-command-and-the-gunzip-command" class="md-nav__link">
    The gzip command and the gunzip command
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-zcat-command" class="md-nav__link">
    The zcat command
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reading-fastq-files" class="md-nav__link">
    Reading FASTQ files
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fastq-format" class="md-nav__link">
    FASTQ Format
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#counting-the-number-of-molecules-reads" class="md-nav__link">
    Counting the number of molecules (reads)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#finding-an-exact-sequence" class="md-nav__link">
    Finding an exact sequence
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#enumerating-lines-of-a-fastq-file" class="md-nav__link">
    Enumerating lines of a FASTQ file
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary" class="md-nav__link">
    Summary
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/sbooeshaghi/bytesizebio/edit/master/docs/tutorials/terminal_interacting_with_fastq_files.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <p><a href="https://colab.research.google.com/github/sbooeshaghi/bytesizebio/blob/main/tutorials/staging/terminal_interacting_with_fastq_files.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h1 id="interacting-with-fastq-files">Interacting with FASTQ files<a class="headerlink" href="#interacting-with-fastq-files" title="Permanent link">&para;</a></h1>
<p>FASTQ files are text files that contain biological sequences, usually nucleotides, and their associated quality scores. The quality score is a measure of how accurate the base-call is for each letter in the sequence. For more information please see the <a href="https://en.wikipedia.org/wiki/FASTQ_format">wikipedia entry</a>.</p>
<hr />
<p><strong>tl;dr</strong> In this tutorial we introduce ways to interact with FASTQ files by the command line.</p>
<h2 id="the-curl-command">The <code>curl</code> command<a class="headerlink" href="#the-curl-command" title="Permanent link">&para;</a></h2>
<p>We will use the <code>curl</code> command to download FASTQ files to our machine. <a href="https://curl.se/"><code>curl</code></a> is a command line tool that allows us to download files. It takes a URL as an argument and downloads the file to your local computer in the current directory. For most cases when downloading data with <code>curl</code> we will use the following syntax:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>$ curl -Ls -o outputfile.txt http://www.somewebsite.com/myfile.txt
</code></pre></div>
</td></tr></table>
<p>Where the <code>-L</code> tells the command to follow "redirects" in case the file as been moved, <code>-s</code> tells the command to be silent and the <code>-o</code> tells the command the name of the file to be saved. </p>
<h2 id="downloading-fastq-files">Downloading FASTQ files<a class="headerlink" href="#downloading-fastq-files" title="Permanent link">&para;</a></h2>
<p>FASTQ files usually come in pairs. They are coloquially called "read 1" and "read 2". This naming refers to the order in which the molecules were sequenced. We won't cover the many ways in which FASTQ files can be generated in this tutorial but for more details please check out these resources: <a href="https://web.archive.org/web/20200414070028/https://www.illumina.com/content/dam/illumina-support/documents/documentation/system_documentation/miseq/indexed-sequencing-overview-guide-15057455-06.pdf">Illumina documentation</a>, <a href="https://ngsc.med.upenn.edu/faqs-mini/FASTQ-Files.html">Next Generation Sequencing Core @UPenn</a>.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!curl -Ls -o read1.fastq https://caltech.box.com/shared/static/fh81mkceb8ydwma3tlrqfgq22z4kc4nt.gz
!curl -Ls -o read2.fastq https://caltech.box.com/shared/static/ycxkluj5my7g3wiwhyq3vhv71mw5gmj5.gz
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!ls -lht
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">total 434M</span>
<span class="err">-rw-r--r-- 1 root root 247M Jan  4 23:49 read2.fastq</span>
<span class="err">-rw-r--r-- 1 root root 188M Jan  4 23:48 read1.fastq</span>
<span class="err">drwxr-xr-x 1 root root 4.0K Dec 21 17:29 sample_data</span>
</code></pre></div>

<h2 id="interacting-with-fastq-files_1">Interacting with FASTQ files<a class="headerlink" href="#interacting-with-fastq-files_1" title="Permanent link">&para;</a></h2>
<p>FASTQ files can be quite large, often on the order of tens to hundreds of gigabytes, therefore it is impractical to open them up and view them in a text editor. For this reason FASTQ files are often "compressed" to save storage space. The type of compression is colloquially called "gzip" compression to refer to the command line tool that compresses files.</p>
<h3 id="useful-commands">Useful commands<a class="headerlink" href="#useful-commands" title="Permanent link">&para;</a></h3>
<h4 id="the-gzip-command-and-the-gunzip-command">The <code>gzip</code> command and the <code>gunzip</code> command<a class="headerlink" href="#the-gzip-command-and-the-gunzip-command" title="Permanent link">&para;</a></h4>
<p>Often times FASTQ files end with <code>.fastq.gz</code>. This means that they have been compressed with the <a href="https://www.gnu.org/software/gzip/"><code>gzip</code></a> command. They can be decompressed them with the <a href="https://linux.die.net/man/1/gunzip"><code>gunzip</code></a> command. Though most of the time we don't want to decompress them since that would require a lot of space. The syntax for decompressing and compressing a file in most cases will look like the following:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>$ gunzip file.fastq.gz 
$ gzip file.fastq
</code></pre></div>
</td></tr></table>
<p>Our files are currently decompressed. Let's see how much space we save by compressing them.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!ls -lht
!gzip read1.fastq
!gzip read2.fastq
!ls -lht
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">total 434M</span>
<span class="err">-rw-r--r-- 1 root root 247M Jan  4 23:49 read2.fastq</span>
<span class="err">-rw-r--r-- 1 root root 188M Jan  4 23:48 read1.fastq</span>
<span class="err">drwxr-xr-x 1 root root 4.0K Dec 21 17:29 sample_data</span>
<span class="err">total 74M</span>
<span class="err">-rw-r--r-- 1 root root  47M Jan  4 23:49 read2.fastq.gz</span>
<span class="err">-rw-r--r-- 1 root root  28M Jan  4 23:48 read1.fastq.gz</span>
<span class="err">drwxr-xr-x 1 root root 4.0K Dec 21 17:29 sample_data</span>
</code></pre></div>

<p>The <code>read1</code> FASTQ file went from 188 Megabytes in size (decompressed, <code>read1.fastq</code>) to 28 Megabytes in size (compressed, <code>read1.fastq.gz</code>) and the <code>read2</code> FASTQ file went from 247 Megabytes (decompressed, <code>read2.fastq</code>) to 47 Megabytes (compressed, <code>read2.fastq.gz</code>). The <code>read1</code> file was compressed by a factor of 6.7 and the <code>read2</code> file was compressed by a factor of 5.2.</p>
<h4 id="the-zcat-command">The <code>zcat</code> command<a class="headerlink" href="#the-zcat-command" title="Permanent link">&para;</a></h4>
<p>Since we almost always want to keep our FASTQ files compressed but we still want to be able to read them we will use a modified version of the <code>cat</code> command but for files that have been compressed with <code>gzip</code>. The <a href="https://linux.die.net/man/1/zcat"><code>zcat</code></a> command does exactly the same thing to a file as the <code>cat</code>, namely it prints the contents of the file to the screen. The difference, however is that the <code>zcat</code> command expects a compressed file. The syntax looks like this:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>$ zcat file.fastq.gz
</code></pre></div>
</td></tr></table>
<h3 id="reading-fastq-files">Reading FASTQ files<a class="headerlink" href="#reading-fastq-files" title="Permanent link">&para;</a></h3>
<p>If we simply use <code>zcat</code> as above it will print the whole file to the screen so instead we "pipe" the output of the zcat command to the <code>head</code> command. We don't conver pipes in this video but there are a lot of great resources covering them. Here is a <a href="https://www.youtube.com/watch?v=bKzonnwoR2I">nice video</a> explaining pipes.</p>
<p>The <code>head</code> command just stops the process of printing after a certain number of lines. The default is 10 lines.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!zcat read1.fastq.gz | head
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">@SRR8611943.1 NS500272:478:HVL5HBGX5:1:11101:9611:1040 length=26</span>
<span class="err">ACATCNGTCGAGAACGATCGTGTCCG</span>
<span class="err">+SRR8611943.1 NS500272:478:HVL5HBGX5:1:11101:9611:1040 length=26</span>
<span class="err">AAAAA#EEEEEEEEEEEEEEEEEEEE</span>
<span class="err">@SRR8611943.2 NS500272:478:HVL5HBGX5:1:11101:20079:1043 length=26</span>
<span class="err">TCAGCNCCAACTGCTAGTCTTTCCCT</span>
<span class="err">+SRR8611943.2 NS500272:478:HVL5HBGX5:1:11101:20079:1043 length=26</span>
<span class="err">AAAAA#EEEEEE6EEEEEEEEEEEEE</span>
<span class="err">@SRR8611943.3 NS500272:478:HVL5HBGX5:1:11101:10651:1043 length=26</span>
<span class="err">TGAAANATCACGCGGTTCATCAGTAG</span>
</code></pre></div>

<p>We can change the number of lines printed by <code>head</code> by specifying a <code>-n</code> option:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!zcat read1.fastq.gz | head -n 4
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">@SRR8611943.1 NS500272:478:HVL5HBGX5:1:11101:9611:1040 length=26</span>
<span class="err">ACATCNGTCGAGAACGATCGTGTCCG</span>
<span class="err">+SRR8611943.1 NS500272:478:HVL5HBGX5:1:11101:9611:1040 length=26</span>
<span class="err">AAAAA#EEEEEEEEEEEEEEEEEEEE</span>
</code></pre></div>

<h3 id="fastq-format">FASTQ Format<a class="headerlink" href="#fastq-format" title="Permanent link">&para;</a></h3>
<p>Every four lines in a FASTQ file represent one molecule that was sequenced.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>@SRR8611943.1 NS500272:478:HVL5HBGX5:1:11101:9611:1040 length=26
ACATCNGTCGAGAACGATCGTGTCCG
+SRR8611943.1 NS500272:478:HVL5HBGX5:1:11101:9611:1040 length=26
AAAAA#EEEEEEEEEEEEEEEEEEEE
</code></pre></div>
</td></tr></table>
<p>The first line is the molecule identifier, the second line is the actual sequence of the molecule, the third line starts with a <code>+</code> sign and provides optional information, and the fourth line is the quality score of the sequence. Note that the quality score must be the same length of the molecule. To learn more about quality scores please check out the <a href="https://support.illumina.com/help/BaseSpace_OLH_009008/Content/Source/Informatics/BS/QualityScoreEncoding_swBS.htm">Illumina documentation</a>.</p>
<h3 id="counting-the-number-of-molecules-reads">Counting the number of molecules (reads)<a class="headerlink" href="#counting-the-number-of-molecules-reads" title="Permanent link">&para;</a></h3>
<p>A common task for bioinformaticians is to count the number of reads (also known as the number of molecules) in a FASTQ file. This means we need to count the number of lines in the FASTQ file and then divide by four, since a read is represented by four lines.</p>
<p>We will use the <a href="https://linuxize.com/post/linux-wc-command/"><code>wc</code></a> command to count the number of lines in a file. The <code>wc</code> command is a commandline tool to count the number of words, characters, and lines in a file. The syntax for counting lines in a file is the following:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>$ wc -l myfile.txt
</code></pre></div>
</td></tr></table>
<p>We will use the <code>wc</code> command to count the number of lines in the FASTQ file and then we will divide by four to get the number of reads.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!zcat read1.fastq.gz | wc -l
!zcat read2.fastq.gz | wc -l
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">4000000</span>
<span class="err">4000000</span>
</code></pre></div>

<p>Both read1 and read2 have 4,000,000 lines, this means that they have 1 million reads. For "paired-end" sequencing data where read 1 and read 2 are given, both FASTQ files will always have the same number of reads since sequneces from a single molecule are represented in both files.</p>
<h3 id="finding-an-exact-sequence">Finding an exact sequence<a class="headerlink" href="#finding-an-exact-sequence" title="Permanent link">&para;</a></h3>
<p>Another common task for bioinformaticians is to find an exact sequence in the FASTQ files. We will use the <code>grep</code> command along with the <code>zcat</code> command. The <a href="https://man7.org/linux/man-pages/man1/grep.1.html"><code>grep</code></a> command is a commandline tool that enables searching. The <code>grep</code> command can get complicated but most use cases can be covered by the following syntax:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>$ grep &quot;find this&quot; myfile.txt
</code></pre></div>
</td></tr></table>
<p>Let's find the sequence <code>ATTAGGAGCCG</code> in <code>read1.fastq.gz</code>.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!zcat read2.fastq.gz | grep &quot;ATTAGGAGCCG&quot;
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">GAAGATGTTGTCGTGGATACTGAAATGCGTCGTCAAAAATTAGGAGCCGTTCTTTTG</span>
<span class="err">GTCAAAAATTAGGAGCCGTTCTTTTGAAGACTCTTGTTTCTCTTGGAAAGTCTCTCG</span>
<span class="err">GCGTCGTCAAAAATTAGGAGCCGTTGGTTTGAAGAATCTTTTTTCTAATGGAAAGTG</span>
<span class="err">GCGTCGTCAAAAATTAGGAGCCGTTCTTTTGAAGACTCTTGTTTCTCTTGGAAAGTC</span>
<span class="err">CAAAAATTAGGAGCCGTTCTTTTGAAGACTCTTGTTTCTCTTGGAAAGTCTCTCGGA</span>
</code></pre></div>

<h3 id="enumerating-lines-of-a-fastq-file">Enumerating lines of a FASTQ file<a class="headerlink" href="#enumerating-lines-of-a-fastq-file" title="Permanent link">&para;</a></h3>
<p>In the above example we found five reads that contain our sequence of interest but we don't know what line number they came from. We can figure this out by using the <code>nl</code> command. The <code>nl</code> command is a commandline tool to add line numbers to the file. The syntax is as follows:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>$ nl file.txt
</code></pre></div>
</td></tr></table>
<p>Let's find out the line numbers where we found our sequence.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!zcat read2.fastq.gz | nl | grep &quot;ATTAGGAGCCG&quot;
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">535046  GAAGATGTTGTCGTGGATACTGAAATGCGTCGTCAAAAATTAGGAGCCGTTCTTTTG</span>
<span class="err">745522  GTCAAAAATTAGGAGCCGTTCTTTTGAAGACTCTTGTTTCTCTTGGAAAGTCTCTCG</span>
<span class="err">818990  GCGTCGTCAAAAATTAGGAGCCGTTGGTTTGAAGAATCTTTTTTCTAATGGAAAGTG</span>
<span class="err">1654078 GCGTCGTCAAAAATTAGGAGCCGTTCTTTTGAAGACTCTTGTTTCTCTTGGAAAGTC</span>
<span class="err">1956810 CAAAAATTAGGAGCCGTTCTTTTGAAGACTCTTGTTTCTCTTGGAAAGTCTCTCGGA</span>
</code></pre></div>

<h2 id="summary">Summary<a class="headerlink" href="#summary" title="Permanent link">&para;</a></h2>
<p>In this tutorial we learned how to interact with FASTQ files using built in shell commands to perform simple tasks. </p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>
</code></pre></div>
</td></tr></table>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../.." class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Home
              </div>
            </div>
          </a>
        
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2020 - 2020 A. Sina Booeshaghi
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://github.com/sbooeshaghi" target="_blank" rel="noopener" title="github.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
      </a>
    
      
      
        
        
      
      <a href="https://twitter.com/sinabooeshaghi" target="_blank" rel="noopener" title="twitter.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/vendor.08c56446.min.js"></script>
      <script src="../../assets/javascripts/bundle.6ced434e.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "../..",
          features: ['search.suggest', 'search.highlight'],
          search: Object.assign({
            worker: "../../assets/javascripts/worker/search.8c7e0a7e.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>